### PR TITLE
New C Cheatsheet

### PR DESCRIPTION
**Pull request by harisphnx**
**What does your Instant Answer do?**
Its a cheatsheet for the Basics of the C programming language.

**What problem does your Instant Answer solve (Why is it better than organic links)?**
It gives access to some of the basic of the C language like conversion characters, and precedence order.

**What is the data source for your Instant Answer? (Provide a link if possible)**
http://www.dummies.com/how-to/content/beginning-c-programming-for-dummies-cheat-sheet.html

**Why did you choose this data source?**
Because this contains a good collection of basic c content

**Are there any other alternative (better) data sources?**
I didn't find any

**What are some example queries that trigger this Instant Answer?**
c cheatsheet OR C cheatsheet. I can add more aliases if needed.

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
Programmers

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
No

**Which existing Instant Answers will this one supercede/overlap with?**
I guess a one which talks about c programming language, i didn't refer any though.

**Are you having any problems? Do you need our help with anything?**
No, Just need it reviewed

**Where did you hear about DuckDuckHack? (For first time contributors)**
The internet

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**

screenshot from 2015-10-28 23 38 52

-----
IA Page: https://duck.co/ia/view/c_cheat_sheet

### CODE REVIEW

**Comment by daxtheduck** [comment]
@harisphnx Hey!
        
Thanks for taking the time to contribute! We really appreciate it.

We work closely with every contributor to make Instant Answers the best they can be, so we appreciate your patience as we look over your code. From here, the process usually goes like this:

1. Pull Request is reviewed by the DuckDuckGo staff and community
2. Staff and community will leave feedback with any necessary updates to the function or design.
3. Once you've made any necessary corrections, then your Instant Answer will be merged and deployed live on DuckDuckGo!

If you have any questions along the way, feel free to ask them here. Our staff and community are also available on Slack to answer any questions you may have. If you'd like to join us there please send an email to <quackslack@duckduckgo.com>.

>More Info: https://duck.co/duckduckhack/submission_and_review

Thanks!

File: share/goodie/cheat_sheets/json/c.json [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    "id": "c_cheat_sheet",
+    "name": "C Cheat Sheet",
+    "description": "Computer Programming Language",
+    "metadata":
+    {
+        "sourceName": "C for Beginners",
+        "sourceUrl": "http://www.dummies.com/how-to/content/beginning-c-programming-for-dummies-cheat-sheet.html"
+    },
+    "aliases": [
+        "C"
```
Comment (moollaza) [inline share/goodie/cheat_sheets/json/c.json]: This alias isn't necessary, by default `c` is already a trigger, because of the filename `c.json`. Aliases are for other, additional triggers. For example "c language"

File: share/goodie/cheat_sheets/json/c.json [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    ],
+    "template_type": "code",
+    "section_order": [
+                      "Comparison Operators",
```
Comment (moollaza) [inline share/goodie/cheat_sheets/json/c.json]: nit: indentation is broken here

File: share/goodie/cheat_sheets/json/c.json [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+            "key": "\\0"
+        }
+     ],
+    "Order of Precedence": [
+        {
+            "val": "Unary - Logical not; associativity goes right to left",
+            "key": "!"
+        }, {
+            "val": "Unary - Increment, decrement; associativity goes from right to left",
+            "key": "[++] , [--]"
```
Comment (moollaza) [inline share/goodie/cheat_sheets/json/c.json]: let's remove the space before the comma `,` in this and the following lines

File: share/goodie/cheat_sheets/json/c.json [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+            "key": "\\\""
+        }, {
+            "val": "The question mark",
+            "key": "\\?"
+        }, {
+            "val": "The “null” byte (backslash-zero)",
+            "key": "\\0"
+        }
+     ],
+    "Order of Precedence": [
```
Comment (moollaza) [inline share/goodie/cheat_sheets/json/c.json]: I'm a little torn, but I think we should break this up into several sections. I'd like to break up these sub-sets:

The wiki article on operators has a few tables that group these. I think we should organize them in the same way: https://en.wikipedia.org/wiki/Operators_in_C_and_C%2B%2B

File: share/goodie/cheat_sheets/json/c.json (hunk 1 above)
Comment (harisphnx) [inline share/goodie/cheat_sheets/json/c.json]: Should i add the alias c language then? Will it work with the spaces inbetween..

File: share/goodie/cheat_sheets/json/c.json (hunk 2 above)
Comment (harisphnx) [inline share/goodie/cheat_sheets/json/c.json]: Will correct this.

File: share/goodie/cheat_sheets/json/c.json (hunk 3 above)
Comment (harisphnx) [inline share/goodie/cheat_sheets/json/c.json]: Are you talking about the comma in "order of precedence unary operator??

File: share/goodie/cheat_sheets/json/c.json (hunk 4 above)
Comment (harisphnx) [inline share/goodie/cheat_sheets/json/c.json]: I am not sure that would be good. The tables that are dividing these operators are on the basis of their function. Here it is showing the precedence of those operators. There is another table called "Operator precedence" on the same page. Its somewhat like that.

File: share/goodie/cheat_sheets/json/c.json (hunk 3 above)
Comment (moollaza) [inline share/goodie/cheat_sheets/json/c.json]: yes.

`"key": "[++] , [--]" ` should be `"key": "[++], [--]" `

File: share/goodie/cheat_sheets/json/c.json (hunk 4 above)
Comment (moollaza) [inline share/goodie/cheat_sheets/json/c.json]: I think we should explain the order of precedence separately then, or we should remove the "sub-category" from the descriptions because it feels like a little too much text/info is going on in this single section.

We could have another section Order of precedence, and mention the name of the groups, like this: http://www.swansontec.com/sopc.html

```
Order of Precedence:
    - `!, ++, --` - Unary
     - `*, /, +, -, ...` Math
    - ...
```

Something like that. So we're explaining the operators in one place, and giving the precedence separately.

Comment (harisphnx) [inline share/goodie/cheat_sheets/json/c.json]: Thats a good idea. Will do that. One more thing, I am new to all this, should i make the changes in my local code and do another pull request?

File: share/goodie/cheat_sheets/json/c.json [hunk 5]
```diff
@@ -0,0 +1,282 @@
+{
+    "id": "c_cheat_sheet",
+    "name": "C Cheat Sheet",
+    "description": "Computer Programming Language",
+    "metadata": {
+        "sourceName": "C for Beginners",
+        "sourceUrl": "http://www.dummies.com/how-to/content/beginning-c-programming-for-dummies-cheat-sheet.html"
+    },
+    "aliases": [
+        "c programming",
```
Comment (mintsoft) [inline share/goodie/cheat_sheets/json/c.json]: Nit: Can you fix the indentation here?

Comment (harisphnx) [inline share/goodie/cheat_sheets/json/c.json]: Can you point out where.. pardon me, but i can't seem to make out where.. Also, i don't know what it means, but why was the pull request closed?

**Comment by mintsoft** [comment]
@harisphnx Apologies for the confusion, this PR hasn't been closed; another C CheatSheet has been closed in favour of this one

File: share/goodie/cheat_sheets/json/c.json (hunk 5 above)
Comment (mintsoft) [inline share/goodie/cheat_sheets/json/c.json]: Apologies, I should have been more descriptive. Line 10 is indented with 9 spaces (which should probably be 8) and line 11 is indented with a hard-tab character, we've settled on 4 spaces/tab and soft-tabs (i.e. space characters)

File: share/goodie/cheat_sheets/json/c.json [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+                "key": "a && b"
+            },
+            {
+                "val": "Logical OR",
+                "key": "a || b"
+            }
+        ],
+        "Bitwise Operators": [
+            {
+                "val": "Bitwise NOT",
```
Comment (mintsoft) [inline share/goodie/cheat_sheets/json/c.json]: This is 1's complement, i.e. flip all the bits, I'm not sure if I'd 'get that' from Bitwise NOT

File: share/goodie/cheat_sheets/json/c.json [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+            {
+                "val": "Size-of",
+                "key": "sizeof (a)"
+            },
+            {
+                "val": "Ternary conditional",
+                "key": "a ? b : c"
+            },
+            {
+                "val": "Comma",
```
Comment (mintsoft) [inline share/goodie/cheat_sheets/json/c.json]: ? Is this for `for` loops etc?

File: share/goodie/cheat_sheets/json/c.json (hunk 6 above)
Comment (harisphnx) [inline share/goodie/cheat_sheets/json/c.json]: Thats how i read in many places. Any suggestions what could be better?

File: share/goodie/cheat_sheets/json/c.json (hunk 7 above)
Comment (harisphnx) [inline share/goodie/cheat_sheets/json/c.json]: No, its a ternary operator. https://en.wikipedia.org/wiki/%3F:#C

File: share/goodie/cheat_sheets/json/c.json (hunk 5 above)
Comment (harisphnx) [inline share/goodie/cheat_sheets/json/c.json]: corrected

File: share/goodie/cheat_sheets/json/c.json [hunk 8]
```diff
@@ -0,0 +1,282 @@
+{
+    "id": "c_cheat_sheet",
+    "name": "C Cheat Sheet",
```
Comment (moollaza) [inline share/goodie/cheat_sheets/json/c.json]: The `name` shouldn't contain `Cheat Sheet` -- let's remove it.

File: share/goodie/cheat_sheets/json/c.json [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+    "description": "Computer Programming Language",
```
Comment (moollaza) [inline share/goodie/cheat_sheets/json/c.json]: Let's describe the contents of the Cheat Sheet instead

File: share/goodie/cheat_sheets/json/c.json (hunk 4 above)
Comment (moollaza) [inline share/goodie/cheat_sheets/json/c.json]: I see you figured it out, but updates should be made locally, and pushed to your Git Branch. Then the PR will update automatically

File: share/goodie/cheat_sheets/json/c.json (hunk 8 above)
Comment (harisphnx) [inline share/goodie/cheat_sheets/json/c.json]: Should i just keep "C" then?

File: share/goodie/cheat_sheets/json/c.json (hunk 9 above)
Comment (harisphnx) [inline share/goodie/cheat_sheets/json/c.json]: Something like, "Basic C operators and more" for the "description" tab

Comment (moollaza) [inline share/goodie/cheat_sheets/json/c.json]: Yes. "Overview of basic operators, order of precedence, and ..."

File: share/goodie/cheat_sheets/json/c.json (hunk 8 above)
Comment (moollaza) [inline share/goodie/cheat_sheets/json/c.json]: It looks like GoLang has a title of "Go Language" I think "C Language" works?

File: share/goodie/cheat_sheets/json/c.json (hunk 6 above)
Comment (mintsoft) [inline share/goodie/cheat_sheets/json/c.json]: Hmm, the only names I know are `1's complement` or `invert bitmask`; @moollaza thoughts?

File: share/goodie/cheat_sheets/json/c.json (hunk 7 above)
Comment (mintsoft) [inline share/goodie/cheat_sheets/json/c.json]: Sorry, the question mark was to indicate my confusion, not asking what the symbol is for. The only case in C where I can think that Comma is used is in for loops `for(int x=0, y=5; x++, y--; x>0)` and params? It seems like an odd operator to document

File: share/goodie/cheat_sheets/json/c.json (hunk 8 above)
Comment (harisphnx) [inline share/goodie/cheat_sheets/json/c.json]: The "name" is suppose to be the name of the instant answer or the name for the topic?

File: share/goodie/cheat_sheets/json/c.json (hunk 9 above)
Comment (harisphnx) [inline share/goodie/cheat_sheets/json/c.json]: "Overview of Basic Operators, Order of Precedence, Escape Sequences and Conversion Characters" 

How is this? Is it too big for the description?

File: share/goodie/cheat_sheets/json/c.json (hunk 6 above)
Comment (harisphnx) [inline share/goodie/cheat_sheets/json/c.json]: I can add something like "Bitwise NOT / one's complement"

File: share/goodie/cheat_sheets/json/c.json (hunk 7 above)
Comment (harisphnx) [inline share/goodie/cheat_sheets/json/c.json]: It has some more uses. See this --> https://en.wikipedia.org/wiki/Comma_operator#Uses

File: share/goodie/cheat_sheets/json/c.json (hunk 6 above)
Comment (jophab) [inline share/goodie/cheat_sheets/json/c.json]: @harisphnx Me too think 1's complement is good

File: share/goodie/cheat_sheets/json/c.json (hunk 7 above)
Comment (jophab) [inline share/goodie/cheat_sheets/json/c.json]: @harisphnx You are right that comma oprerator has more uses. But when we look at the cheat sheet  and see
        
               "val": "Comma",
                "key": "a, b"
        it makes confusion since the val field cant specify what it does..

File: share/goodie/cheat_sheets/json/c.json (hunk 8 above)
Comment (mintsoft) [inline share/goodie/cheat_sheets/json/c.json]: @harisphnx for the cheatsheet's it's usually the topic

File: share/goodie/cheat_sheets/json/c.json (hunk 9 above)
Comment (mintsoft) [inline share/goodie/cheat_sheets/json/c.json]: @harisphnx that should be OK, the description isn't really limited length wise

File: share/goodie/cheat_sheets/json/c.json (hunk 6 above)
Comment (harisphnx) [inline share/goodie/cheat_sheets/json/c.json]: @mintsoft @jophab Corrected. I have changed it to "Bitwise NOT / One's Complement"

File: share/goodie/cheat_sheets/json/c.json (hunk 8 above)
Comment (harisphnx) [inline share/goodie/cheat_sheets/json/c.json]: @mintsoft @moollaza Changed it to "C Language".

File: share/goodie/cheat_sheets/json/c.json (hunk 9 above)
Comment (harisphnx) [inline share/goodie/cheat_sheets/json/c.json]: @mintsoft @moollaza  Changed it to "Overview of Basic Operators, Order of Precedence, Escape Sequences and Conversion Characters".

File: share/goodie/cheat_sheets/json/c.json (hunk 7 above)
Comment (harisphnx) [inline share/goodie/cheat_sheets/json/c.json]: @jophab @mintsoft The problem is comma operator has distinct uses for distinct things. For expressions it works in a different way, for function parameters differently and for "for" loops differently. I cannot come up with anything that can define the comma operator in one line.

Comment (harisphnx) [inline share/goodie/cheat_sheets/json/c.json]: @jophab @mintsoft I am thinking of dropping the comma operator. suggestion?

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=c cheat sheet)
                
                [Mobile screenshot](https://ia-screenshots.s3.amazonaws.com/c_cheat_sheet_mobile.png)

File: share/goodie/cheat_sheets/json/c.json (hunk 7 above)
Comment (mintsoft) [inline share/goodie/cheat_sheets/json/c.json]: @harisphnx I'd support dropping it

Comment (harisphnx) [inline share/goodie/cheat_sheets/json/c.json]: @jophab Your call?

Comment (jophab) [inline share/goodie/cheat_sheets/json/c.json]: @harisphnx Me too support droping :+1:

Comment (harisphnx) [inline share/goodie/cheat_sheets/json/c.json]: @jophab @mintsoft Removed.

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=c cheat sheet)
                
                [Mobile screenshot](https://ia-screenshots.s3.amazonaws.com/c_cheat_sheet_mobile.png)

**Comment by mintsoft** [comment]
Apologies for the delay getting back to you; this looks good to me. Thanks for the contribution @harisphnx

**Comment by jophab** [comment]
@harisphnx congrats... :+1:

**Comment by harisphnx** [comment]
@mintsoft you're welcome. And now I can start work on the gdb cheatsheet. :)

**Comment by harisphnx** [comment]
@jophab Thanks :)
